src/node/ssh.ts

```diff
@@ -0,0 +1,293 @@
+import { IProcessEnvironment } from 'vs/base/common/platform';
+
+const ssh = localRequire<typeof import("ssh2")>("ssh2/lib/client");
+
+/**
+ * Fills out all of the functionality of SSH using node equivalents.
+ */
+export function fillSSHSession(accept: () => sshTypes.Session) {
+	let pty: nodePty.IPty | undefined;
+	let activeProcess: cp.ChildProcess;
+	let ptyInfo: sshTypes.PseudoTtyInfo | undefined;
+	const env: { [key: string]: string } = {};
+
+	const session = accept();
+
+	// Run a command, stream back the data
+	const cmd = (command: string, channel: sshTypes.ServerChannel): void => {
+		if (ptyInfo) {
+			// Remove undefined env vars and remove VSCode / code-server env vars
+			const env = Object.keys(process.env).reduce((prev, k) => {
+				if (process.env[k] !== undefined) {
+					prev[k] = process.env[k] as string;
+				}
+				return prev;
+			}, {} as IProcessEnvironment);
+			sanitizeProcessEnvironment(env as IProcessEnvironment);
+
+			pty = nodePty.spawn(command, [], {
+				cols: ptyInfo.cols,
+				rows: ptyInfo.rows,
+				env,
+			});
+			pty.onData((d) => channel.write(d));
+			pty.on("exit", (exitCode) => {
+				channel.exit(exitCode);
+				channel.close();
+			});
+			channel.on("data", (d: any) => pty!.write(d));
+			return;
+		}
+
+		const proc = cp.spawn(command);
+		proc.stdout.on("data", (d) => channel.stdout.write(d));
+		proc.stderr.on("data", (d) => channel.stderr.write(d));
+		proc.on("exit", (exitCode) => {
+				channel.exit(exitCode || 0);
+				channel.close();
+		});
+		channel.stdin.on("data", (d: any) => proc.stdin.write(d));
+		channel.stdin.on("close", () => proc.stdin.end());
+	};
+
+	session.on("pty", (accept, _, info) => {
+		ptyInfo = info;
+		accept();
+	});
+
+	session.on("shell", accept => {
+		cmd(process.env.SHELL || "/usr/bin/env bash", accept());
+	});
+
+	session.on("exec", (accept, _, info) => {
+		cmd(info.command, accept());
+	});
+
+	session.on("sftp", fillSFTPStream);
+
+	session.on("signal", (accept, _, info) => {
+		accept();
+		process.kill((pty || activeProcess).pid, info.name);
+	});
+
+	session.on("env", (_accept, _reject, info) => {
+		env[info.key] = info.value;
+	});
+
+	session.on("window-change", (accept, reject, info) => {
+		if (pty) {
+			pty.resize(info.cols, info.rows);
+			accept();
+		} else {
+			reject();
+		}
+	});
+}
+
+/**
+ * Fills out all the functionality of SFTP using fs.
+ */
+function fillSFTPStream(accept: () => SFTPStream) {
+	const sftp = accept();
+
+	let oid = 0;
+	const fds: { [key: number]: boolean } = {};
+	const ods: { [key: number]: {
+		path: string
+		read: boolean
+	} } = {};
+
+	const sftpStatus = (reqID: number, err?: NodeJS.ErrnoException) => {
+		let code = ssh.SFTP_STATUS_CODE.OK;
+		if (err) {
+			if (err.code === "EACCES") {
+				code = ssh.SFTP_STATUS_CODE.PERMISSION_DENIED;
+			}
+			if (err.code === "ENOENT") {
+				code = ssh.SFTP_STATUS_CODE.NO_SUCH_FILE;
+			}
+			code = ssh.SFTP_STATUS_CODE.FAILURE;
```

Comment:
`code` will always be `FAILURE`.

Reply:
Jeez, I was really asleep at the wheel on this one 😴